src/transaction/log_recovery_redo_parallel.hpp

```diff
@@ -299,6 +315,43 @@ namespace cublog
       const log_reader::fetch_mode m_log_reader_page_fetch_mode;
   };
 
+
+  /* job implementation that performs log replication delay calculation
+   * using log records that register creation time
+   */
+  template <typename TYPE_LOG_REC> // TODO: does not need to be a template
+  class redo_job_replication_delay_impl final : public redo_parallel::redo_job_base
+  {
+      using log_rec_t = TYPE_LOG_REC;
+
+      /* sentinel VPID value needed for the internal mechanics of the parallel log recovery/replication
+       * internally, such a VPID is needed to maintain absolute order of the processing
+       * of the log records with respect to their order in the global log record
+       */
+      static constexpr short SENTINEL_VOLID = -2;
+      static constexpr int32_t SENTINEL_PAGEID = -2;
+      static constexpr vpid SENTINEL_VPID = { SENTINEL_PAGEID, SENTINEL_VOLID };
+
+    public:
+      redo_job_replication_delay_impl (const log_lsa &a_rcv_lsa, time_t a_start_time_msec, const char *a_source);
+
+      redo_job_replication_delay_impl (redo_job_replication_delay_impl const &) = delete;
+      redo_job_replication_delay_impl (redo_job_replication_delay_impl &&) = delete;
+
+      ~redo_job_replication_delay_impl () override = default;
+
+      redo_job_replication_delay_impl &operator = (redo_job_replication_delay_impl const &) = delete;
+      redo_job_replication_delay_impl &operator = (redo_job_replication_delay_impl &&) = delete;
+
+      int execute (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+		   LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support) override;
+
+    private:
+      const time_t m_start_time_msec;
+      const char *const m_source;
```

Comment:
Is source relevant? Can it be removed?
```suggestion
```

Reply:
I think it is.
I noticed that the COMMIT/ABORT delays are quite homogenous. But the HA_REPL one is wildly varying - sometimes similar to COMMIT/ABORT, othertimes much higher values, in the seconds.
TODO: I'll provide a logging snapshot.